common/src/main/java/org/mcaccess/minecraftaccess/features/point_of_interest/POIBlocks.java

```diff
@@ -88,63 +92,91 @@ public class POIBlocks {
-
-    public Map<String, POIGroup> builtInGroups = Map.of(
-        "marked", new POIGroup("Marked blocks", SoundEvents.ENTITY_ITEM_PICKUP, -5f, null, (block, pos) -> onPOIMarkingNow && markedBlock.test(block)),
-    "ore", new POIGroup("Ores", SoundEvents.ENTITY_ITEM_PICKUP, -5f, null, (block, pos) -> oreBlockPredicates.stream().anyMatch(p -> p.test(block))),
-        "fluid", new POIGroup("Fluids", SoundEvents.BLOCK_NOTE_BLOCK_BIT.value(), 2f, null, (block, pos) -> this.detectFluidBlocks && block.getBlock() instanceof FluidBlock && PlayerUtils.isNotInFluid() && block.getFluidState().getLevel() == 8),
-        "functional", new POIGroup("Functional blocks", SoundEvents.BLOCK_NOTE_BLOCK_BIT.value(), 2f, null, (block, pos) -> block.getBlock() instanceof ButtonBlock || block.getBlock() instanceof LeverBlock || poiBlockPredicates.stream().anyMatch(p -> p.test(block))),
-        "gui", new POIGroup("Blocks with interface", SoundEvents.BLOCK_NOTE_BLOCK_BANJO.value(), 0f, null, (block, pos) -> block.createScreenHandlerFactory(world, pos) != null)
+    private @Nullable Block markedBlock = null;
+    private boolean isMarking = false;
+
+    private final POIGroup<BlockPos> markedGroup = new POIGroup<>(
+            SoundEvents.ENTITY_ITEM_PICKUP,
+            -5f,
+            pos -> isMarking && world.getBlockState(pos).isOf(markedBlock)
+    );
+    private final POIGroup<BlockPos> oreGroup = new POIGroup<>(
+            SoundEvents.ENTITY_ITEM_PICKUP,
+            -5f,
+            pos -> oreBlockPredicates.stream().anyMatch(p -> p.test(world.getBlockState(pos)))
     );
 
-    static {
-        try {
-            instance = new POIBlocks();
-        } catch (Exception e) {
-            throw new RuntimeException("Exception occurred in creating POIBlocks instance", e);
-        }
-    }
+    @SuppressWarnings("unchecked")
+    private final POIGroup<BlockPos>[] groups = new POIGroup[] {
+            markedGroup,
+            oreGroup,
+            new POIGroup<BlockPos>(// Fluids
+                    SoundEvents.BLOCK_NOTE_BLOCK_BIT.value(),
+                    2f,
+                    pos -> this.detectFluidBlocks && world.getBlockState(pos).getBlock() instanceof FluidBlock && PlayerUtils.isNotInFluid() && world.getFluidState(pos).getLevel() == 8
+            ),
+            new POIGroup<BlockPos>(// Functional blocks
+                    SoundEvents.BLOCK_NOTE_BLOCK_BIT.value(),
+                    2f,
+                    pos -> world.getBlockState(pos).getBlock() instanceof ButtonBlock || world.getBlockState(pos).getBlock() instanceof LeverBlock || poiBlockPredicates.stream().anyMatch(p -> p.test(world.getBlockState(pos)))
+            ),
+            new POIGroup<BlockPos>(// Blocks with interface
+                    SoundEvents.BLOCK_NOTE_BLOCK_BANJO.value(),
+                    0f,
+                    pos -> world.getBlockState(pos).createScreenHandlerFactory(world, pos) != null
+            ),
+    };
```

Comment:
How about extracting these `POIGroup` instances into their own variable (inside the static block)? To make the meaning of each group more clear.

Reply:
Well, as long as the assignment is inside the static block, I can tell it's sort of initialization, not the runtime result.